chain/src/main/scala/org/bitcoins/chain/blockchain/sync/FilterSync.scala

```diff
@@ -0,0 +1,139 @@
+  * }}}
+  * which will allow us to sync our internal filters against.
+  * */
+abstract class FilterSync extends ChainVerificationLogger {
+
+  def syncFilters(
+      chainApi: ChainApi,
+      getFilterFunc: BlockHeader => Future[GolombFilter])(
+      implicit ec: ExecutionContext,
+      chainAppConfig: ChainAppConfig): Future[ChainApi] = {
+
+    val ourBestFilterHeaderF = chainApi.getBestFilterHeader()
+    val ourBestBlockHeaderF = chainApi.getBestBlockHeader()
+    for {
+      ours <- ourBestFilterHeaderF
+      ourBestBlockHeader <- ourBestBlockHeaderF
+      syncedChainApi <- syncFiltersToTip(chainApi = chainApi,
+                                         ourBestHeader = ourBestBlockHeader,
+                                         ourBestFilterHeader = ours,
+                                         getFilterFunc = getFilterFunc)
+    } yield {
+      syncedChainApi
+    }
+  }
+
+  private case class BlockFilterAggregated(
+      filterHeader: FilterHeader,
+      filter: GolombFilter,
+      blockHeader: BlockHeader)
+
+  /**
+    * Syncs our best filter header to our best block hash
+    * @param chainApi our current chain state
+    * @param ourBestHeader the block header we are going to sync filters up until
+    * @param ourBestFilterHeader the best filter header we have
+    * @param getFilterFunc given a block hash it retrieves filter associated with that hash from our external source
+    * @param ec
+    * @return
+    */
+  private def syncFiltersToTip(
+      chainApi: ChainApi,
+      ourBestHeader: BlockHeaderDb,
+      ourBestFilterHeader: CompactFilterHeaderDb,
+      getFilterFunc: BlockHeader => Future[GolombFilter])(
+      implicit ec: ExecutionContext,
+      chainAppConfig: ChainAppConfig): Future[ChainApi] = {
+    if (ourBestFilterHeader.blockHashBE == ourBestHeader.hashBE) {
+      logger.info(
+        s"Our filters are synced with our peers filters, both at blockHash=${ourBestFilterHeader.blockHashBE}")
+      Future.successful(chainApi)
+    } else {
+      logger.info(
+        s"Beginning sync for filters from filterheader=${ourBestFilterHeader} to blockheader=${ourBestHeader.hashBE}")
+      //let's fetch all missing filter headers first
+      val bestFilterBlockHeaderF =
+        chainApi.getHeader(ourBestFilterHeader.blockHashBE)
+
+      val headersMissingFiltersF = for {
+        bestFilterBlockHeader <- bestFilterBlockHeaderF
+        missing <- chainApi.getHeadersBetween(from = bestFilterBlockHeader.get,
+                                              to = ourBestHeader)
+      } yield missing
+
+      //now that we have headers that are missing filters, let's fetch the filters
+      val fetchFiltersF: Future[Vector[(BlockHeaderDb, GolombFilter)]] = for {
+        missingHeaders <- headersMissingFiltersF
+        fetchNested = missingHeaders.map { b =>
+          val filterF = getFilterFunc(b.blockHeader)
+          filterF.map(f => (b, f))
```

Comment:
Filters can be pretty big, especially on mainnet. Storing all of them in the memory can cause memory issues.